vortex-array/src/arrays/fixed_size_list/compute/filter.rs

```diff
@@ -0,0 +1,151 @@
+            // - `new_validity` has the correct length because we filter with the same
+            //   `selection_mask` as the array itself
+            unsafe {
+                FixedSizeListArray::new_unchecked(new_elements, list_size, new_validity, new_len)
+            }
+            .into_array(),
+        )
+    }
+}
+
+register_kernel!(FilterKernelAdapter(FixedSizeListVTable).lift());
+
+/// Given a mask for a fixed-size list array, creates a new mask for the underlying elements.
+///
+/// This function simply "expands" out the input `selection_mask` by duplicating each bit
+/// `list_size` times.
+///
+/// The output `Mask` is guaranteed to have a length equal to `selection_mask.len() * list_size`.
+fn compute_fsl_elements_mask(selection_mask: &Mask, list_size: usize) -> Mask {
+    let expanded_len = selection_mask.len() * list_size;
+
+    match selection_mask {
+        Mask::AllTrue(_) => Mask::AllTrue(expanded_len),
+        Mask::AllFalse(_) => Mask::AllFalse(expanded_len),
+        Mask::Values(values) => {
+            // Use threshold_iter to choose the optimal representation based on density.
+            match values.threshold_iter(FSL_MASK_EXPANSION_DENSITY_THRESHOLD) {
+                MaskIter::Slices(slices) => expand_dense_mask(slices, list_size, expanded_len),
+                MaskIter::Indices(indices) => expand_sparse_mask(indices, list_size, expanded_len),
+            }
+        }
+    }
+}
+
+/// Expands a dense mask (represented as slices) by scaling each slice by `list_size`.
+fn expand_dense_mask(slices: &[(usize, usize)], list_size: usize, expanded_len: usize) -> Mask {
+    let expanded_slices: Vec<(usize, usize)> = slices
+        .iter()
+        .map(|&(start, end)| (start * list_size, end * list_size))
+        .collect();
+
+    Mask::from_slices(expanded_len, expanded_slices)
+}
+
+/// Expands a sparse mask (represented as indices) by duplicating each index `list_size` times.
+fn expand_sparse_mask(indices: &[usize], list_size: usize, expanded_len: usize) -> Mask {
+    if list_size <= FSL_SPARSE_MASK_LIST_SIZE_THRESHOLD {
+        // For small list sizes, expand each index into individual indices.
+        let expanded_indices: Vec<usize> = indices
+            .iter()
+            .flat_map(|&idx| {
+                let start = idx * list_size;
+                start..start + list_size
+            })
+            .collect();
+
+        Mask::from_indices(expanded_len, expanded_indices)
+    } else {
+        // For sparse masks with large list sizes, it's more efficient to create slices rather than
+        // materializing all individual indices.
+        let expanded_slices: Vec<(usize, usize)> = indices
+            .iter()
+            .map(|&idx| {
+                let start = idx * list_size;
+                let end = (idx + 1) * list_size;
+                (start, end)
+            })
+            .collect();
+
+        Mask::from_slices(expanded_len, expanded_slices)
+    }
```

Comment:
I feel this is uncessarily optimized, we can come back to further optimisation once we have it wired into the file layout